File: 程序员面试金典/No_01_06.py
class Solution:
    def compressString(self, S: str) -> str:
        i = j = 0
        n = len(S)
        res = ''
        cnt = 0
        while i < n:
            if j == n:
                res += str(j - i)
                break
            if i == j:
                res += S[i]
                j += 1
            elif S[i] == S[j]:
                j += 1
            else:
                res += str(j - i)
                i = j
        if len(res) < n:
            return res
        else:
            return S

File: 程序员面试金典/No_01_06.cpp
class Solution {
public:
    string compressString(string S) {
        int cnt = 0;
        string res;
        for (int i = 0; i < S.size(); ++i){
            if (i == 0) ++cnt;
            if (i > 0){
                if(S[i] == S[i - 1]) ++cnt;
                else{
                    res += S[i - 1] + to_string(cnt);
                    cnt = 1;
                }
            }
            if(i == S.size() - 1) res += S[i] + to_string(cnt);
            }
        return res.size() < S.size()? res: S;
    }
};